File: day16/puzzle1.py
from collections import deque
from enum import Enum


class Direction(Enum):
    """Possible directions."""

    RIGHT = 1
    DOWN = 2
    LEFT = 3
    UP = 4


def calculate_energized_tiles(contraption: list[list[str]], start: tuple[int, int, Direction]) -> int:
    """Calculates energized tiles.

    Parameters
    ----------
    contraption : list[list[str]]
        The contraption.
    start : tuple[int, int, Direction]
        The starting point.

    Returns
    -------
    int
        The energized tile count.

    Raises
    ------
    ValueError
        If an invalid tile is encountered.
    """
    legs: set[tuple[int, int, Direction]] = set()
    energized: list[list[bool]] = [[False for _ in row] for row in contraption]
    queue = deque([start])
    height = len(contraption)
    width = len(contraption[0])
    while queue:
        leg = queue.popleft()
        if leg in legs:
            continue
        legs.add(leg)
        print(leg)
        (y, x, direction) = leg
        while True:
            energized[y][x] = True
            tile = contraption[y][x]
            if direction == Direction.RIGHT:
                if tile in ".-":
                    x += 1
                    if x == width:
                        break
                else:
                    if y > 0 and tile in "|/":
                        queue.append((y - 1, x, Direction.UP))
                    if y + 1 < height and tile in "|\\":
                        queue.append((y + 1, x, Direction.DOWN))
                    break
            elif direction == Direction.DOWN:
                if tile in ".|":
                    y += 1
                    if y == height:
                        break
                else:
                    if x > 0 and tile in "-/":
                        queue.append((y, x - 1, Direction.LEFT))
                    if x + 1 < width and tile in "-\\":
                        queue.append((y, x + 1, Direction.RIGHT))
                    break
            elif direction == Direction.LEFT:
                if tile in ".-":
                    x -= 1
                    if x < 0:
                        break
                else:
                    if y > 0 and tile in "|\\":
                        queue.append((y - 1, x, Direction.UP))
                    if y < height - 1 and tile in "|/":
                        queue.append((y + 1, x, Direction.DOWN))
                    break
            else:  # Direction.UP
                if tile in ".|":
                    y -= 1
                    if y < 0:
                        break
                else:
                    if x > 0 and tile in "-\\":
                        queue.append((y, x - 1, Direction.LEFT))
                    if x + 1 < width and tile in "-/":
                        queue.append((y, x + 1, Direction.RIGHT))
                    break
    return sum([sum([0 if not tile else 1 for tile in row]) for row in energized])


def main():
    """Application entry-point."""
    contraption: list[list[str]] = []
    with open("puzzle_input.txt", "rt") as f:
        for line in f:
            contraption.append(list(line.strip()))
    tile_count = calculate_energized_tiles(contraption, (0, 0, Direction.RIGHT))
    print(f"energized tiles = {tile_count}")


if __name__ == "__main__":
    main()

File: day16/puzzle1.dart
import 'dart:collection';
import 'dart:io';
import 'dart:convert';

enum Direction { right, down, left, up }

typedef Leg = (int, int, Direction);

int calculateEnergizedTiles(List<String> contraption, Leg start) {
  var legs = <Leg>{};
  final height = contraption.length;
  final width = contraption[0].length;
  var energized =
      List.generate(height, (_) => List.generate(width, (_) => false));
  var queue = Queue<Leg>();
  queue.addLast(start);
  while (queue.isNotEmpty) {
    final leg = queue.removeFirst();
    if (legs.contains(leg)) {
      continue;
    }
    legs.add(leg);
    var (y, x, direction) = leg;
    var done = false;
    while (!done) {
      energized[y][x] = true;
      final tile = contraption[y][x];
      switch (direction) {
        case Direction.right:
          if ('.-'.contains(tile)) {
            x++;
            if (x == width) {
              done = true;
            }
          } else {
            if (y > 0 && '|/'.contains(tile)) {
              queue.addLast((y - 1, x, Direction.up));
            }
            if (y + 1 < height && '|\\'.contains(tile)) {
              queue.addLast((y + 1, x, Direction.down));
            }
            done = true;
          }
        case Direction.down:
          if ('.|'.contains(tile)) {
            y++;
            if (y == height) {
              done = true;
            }
          } else {
            if (x > 0 && '-/'.contains(tile)) {
              queue.addLast((y, x - 1, Direction.left));
            }
            if (x + 1 < width && '-\\'.contains(tile)) {
              queue.addLast((y, x + 1, Direction.right));
            }
            done = true;
          }
        case Direction.left:
          if ('.-'.contains(tile)) {
            x--;
            if (x < 0) {
              done = true;
            }
          } else {
            if (y > 0 && '|\\'.contains(tile)) {
              queue.addLast((y - 1, x, Direction.up));
            }
            if (y + 1 < height && '|/'.contains(tile)) {
              queue.addLast((y + 1, x, Direction.down));
            }
            done = true;
          }
        case Direction.up:
          if ('.|'.contains(tile)) {
            y--;
            if (y < 0) {
              done = true;
            }
          } else {
            if (x > 0 && '-\\'.contains(tile)) {
              queue.addLast((y, x - 1, Direction.left));
            }
            if (x + 1 < width && '-/'.contains(tile)) {
              queue.addLast((y, x + 1, Direction.right));
            }
            done = true;
          }
      }
    }
  }
  var tile_count = 0;
  for (final row in energized) {
    for (final tile in row) {
      tile_count += (tile ? 1 : 0);
    }
  }
  return tile_count;
}

void main() async {
  final lines = utf8.decoder
      .bind(File('puzzle_input.txt').openRead())
      .transform(const LineSplitter());
  var contraption = <String>[];
  await for (final line in lines) {
    contraption.add(line);
  }
  final tile_count =
      calculateEnergizedTiles(contraption, (0, 0, Direction.right));
  print('energized tiles = $tile_count');
}
